datamol/match.py

```diff
@@ -0,0 +1,139 @@
+from rdkit.Chem import AllChem
+from typing import Dict, List, Union
+import networkx as nx
+from networkx.algorithms.isomorphism.vf2userfunc import GraphMatcher
+from loguru import logger
+
+
+def mol_to_nx_graph(mol: AllChem.rdchem.Mol) -> nx.Graph:
+    """
+    Convert a molecule to a networkx graph.
+
+    Parameters:
+        mol: Molecule to convert
+    Returns:
+        A networkx graph containing an `atomic_num` attribute on the nodes
+        and a `weight` attribute on the edges, corresponding to
+        single=1, double=2, triple=3, quadruple=4, quintuple=5, aromatic=1.5.
+    """
+    atomic_numx_property = {
+        ii: {"atomic_num": atom.GetAtomicNum()} for ii, atom in enumerate(mol.GetAtoms())
+    }
+    adj = AllChem.GetAdjacencyMatrix(mol, useBO=True)
+    g = nx.from_numpy_matrix(adj)
+    nx.set_node_attributes(g, atomic_numx_property)
+    return g
+
+
+def match_molecular_graphs(
+    mol1: AllChem.rdchem.Mol,
+    mol2: AllChem.rdchem.Mol,
+    match_atomic_num: bool = True,
+    match_bond_type: bool = True,
+) -> List[Dict[int, int]]:
+    """
+    Match the node indices of 2 molecular graphs, with optional usage of atomic number and edge type.
+
+    Parameters:
+        mol1, mol2: The molecules to match their indices.
+        match_atomic_num: Whether the atomic number should be used when matching the graphs.
+        match_bond_type: Whether the bond type should be used when matching the graphs.
```

Comment:
I think something worth checking too is the behaviour when hydrogen bonds are explicits.

Reply:
Matching will fail if hydrogens are not explicit/implicit in both. I added a note about this.

Also added another Note:
        Explicit hydrogens might lead to too many matches, since for an atom with 2
        hydrogens, they can be re-ordered in any way.